bodhi/server/models.py

```diff
@@ -616,24 +616,143 @@ class RpmPackage(Package):
+    update_id = Column(Integer, ForeignKey('updates.id'))
 
     release = relationship('Release', backref='builds', lazy=False)
 
+    type = Column(Integer, nullable=False)
+    __mapper_args__ = {
+        'polymorphic_on': type,
+        'polymorphic_identity': 0,
+    }
+
+    def get_url(self):
+        """
+        Return a the url to details about this build.
+
+        This method appears to be unused and incorrect.
+
+        Return:
+            str: A URL for this build.
+        """
+        return '/' + self.nvr
+
+    def get_tags(self, koji=None):
+        """
+        Return a list of koji tags for this build.
+
+        Args:
+            koji (bodhi.server.buildsys.Buildsysem or koji.ClientSession): A koji client. Defaults
+                to calling bodhi.server.buildsys.get_session().
+        Return:
+            list: A list of strings of the Koji tags on this Build.
+        """
+        if not koji:
+            koji = buildsys.get_session()
+        return [tag['name'] for tag in koji.listTags(self.nvr)]
+
+    def untag(self, koji, db):
+        """
+        Remove all known tags from this build.
+
+        Args:
+            koji (bodhi.server.buildsys.Buildsysem or koji.ClientSession): A koji client.
+            db (sqlalchemy.orm.session.Session): A database Session.
+        """
+        tag_types, tag_rels = Release.get_tags(db)
+        for tag in self.get_tags():
+            if tag in tag_rels:
+                log.info('Removing %s tag from %s' % (tag, self.nvr))
+                koji.untagBuild(tag, self.nvr)
+
+    def unpush(self, koji):
+        """
+        Move this build back to the candidate tag and remove any pending tags.
+
+        Args:
+            koji (bodhi.server.buildsys.Buildsysem or koji.ClientSession): A koji client.
+        """
+        log.info('Unpushing %s' % self.nvr)
+        release = self.update.release
+        for tag in self.get_tags(koji):
+            if tag == release.pending_signing_tag:
+                log.info('Removing %s tag from %s' % (tag, self.nvr))
+                koji.untagBuild(tag, self.nvr)
+            if tag == release.pending_testing_tag:
+                log.info('Removing %s tag from %s' % (tag, self.nvr))
+                koji.untagBuild(tag, self.nvr)
+            if tag == release.pending_stable_tag:
+                log.info('Removing %s tag from %s' % (tag, self.nvr))
+                koji.untagBuild(tag, self.nvr)
+            elif tag == release.testing_tag:
+                log.info(
+                    'Moving %s from %s to %s' % (
+                        self.nvr, tag, release.candidate_tag))
+                koji.moveBuild(tag, release.candidate_tag, self.nvr)
+
+
+class RpmBuild(Build):
+    """
+    Represents an RPM build.
+
+    Note that this model uses single-table inheritance with its Build superclass.
+
+    Attributes:
+        nvr (unicode): A dash (-) separated string of an RPM's name, version, and release (e.g.
```

Comment:
@jeremycline Actually, I had left it here on purpose so that the meaning of nvr could be defined for this model specifically. The parent class documents it as a koji identifier, but this adds that it's name-version-release for RPMs.